builder/dockerfile/shell/lex.go

```diff
@@ -261,6 +261,31 @@ func (sw *shellWord) processDollar() (string, error) {
 	}
 
 	sw.scanner.Next()
+
+	if sw.scanner.Peek() == ':' || sw.scanner.Peek() == '-' || sw.scanner.Peek() == '+' {
+		// Invalid ${:xx}, ${-xx}, and ${-} cases
+		//
+		// Note: validity of ${-:bbb} and ${-} is inconsistent between implementations.
+		// Treat it as invalid, because we cannot substitute the `-` with
+		// anything useful.
```

Comment:
Why is `$-` special but `$*` `$@` `$#` `$?` `$$` `$0` and `$_` (other such special parameters in POSIX) do not appear to be?

I also think it is a mistake to conflate the existence/behaviour of those special parameters with the entirely unrelated support for the `${parameter:offset}` syntax, which is non-POSIX and hence not supported by dash but is supported by `bash` and whatever the alpine/busybox shell is these days (`ash`?).

Don't be confused into thinking that `${-:offset}` is somehow making `-:` into something special, it is not, it's just two unrelated concepts being used at the same time.

I think you should decide whether or not the `${parameter:offset}` form is valid, irrespective of whether `parameter` is `-` or `foo` or `PATH` and just implement what you chose there.

Then, completely separately, decide which, if any, special parameters are supported. This could be as simple as having no way to set `-` or `@` etc and therefore treating them as undefined variables at this level or could be done by seeing those names, whenever they are used, and rejecting them.

BTW I don't think `${:xx}` is valid anywhere, is it? Because it is missing the parameter name entirely so it is just syntactically invalid.